eventuals/do-all.h

```diff
@@ -67,11 +99,68 @@ struct _DoAll {
                               StoppedException()));
                   if (counter_.fetch_sub(1) == 1) {
                     // You're the last eventual so call the continuation.
-                    k_.Start(std::move(values_));
+                    std::optional<std::exception_ptr> exception = GetExceptionIfExists();
+
+                    if (exception) {
+                      try {
+                        std::rethrow_exception(*exception);
+                      } catch (const StoppedException&) {
+                        k_.Stop();
+                      } catch (...) {
+                        k_.Fail(std::current_exception());
+                      }
+                    }
                   }
                 }));
     }
 
+    std::tuple<
+        std::conditional_t<
+            std::is_void_v<typename Eventuals_::template ValueFrom<void>>,
+            std::monostate,
+            typename Eventuals_::template ValueFrom<void>>...>
+    GetTupleOfValues() {
+      return std::apply(
+          [](auto&... value) {
+            ((CHECK(value.valueless_by_exception() == false)), ...);
```

Comment:
```suggestion
            (CHECK_EQ(0, value.index()), ...);
```
I love that you're checking this! But rather than checking for `valueless_by_exception()` how about we check for the index? That way we are sure the `std::get<0>` will work below and that not only is it not `valueless_by_exception()` but it also doesn't hold a `std::exception_ptr` at index 1.

Reply:
> I love that you're checking this! But rather than checking for `valueless_by_exception()` how about we check for the index? That way we are sure the `std::get<0>` will work below and that not only is it not `valueless_by_exception()` but it also doesn't hold a `std::exception_ptr` at index 1.

Great idea, Ben, I like it. But the following suggestion - `((CHECK_EQ(0, value.index())), ...);` will not compile. On the screen below you can watch the error. So my suggestion will be to use `CHECK`. That's why the final synax would be: `((CHECK(value.index() == 0)), ...);`
![Снимок экрана от 2021-11-05 16-11-16](https://user-images.githubusercontent.com/68085133/140515706-4d807726-2d8d-4024-b363-42cf2fb1e337.png)